From 97ad1a4051e52f0f58f8f49f74317bcf25e57b39 Mon Sep 17 00:00:00 2001
From: Nick Clifton <nickc@redhat.com>
Date: Tue, 4 Nov 2014 13:15:37 +0000
Subject: [PATCH] Fix a seg-fault triggered by reading a mal-formed archive.

	PR binutils/17533
	* archive.c (_bfd_slurp_extended_name_table): Handle archives with
	corrupt extended name tables.

Signed-off-by: Yuanjie Huang <Yuanjie.Huang@windriver.com>
---
 bfd/ChangeLog | 6 ++++++
 bfd/archive.c | 9 +++++++--
 2 files changed, 13 insertions(+), 2 deletions(-)

diff --git a/bfd/ChangeLog b/bfd/ChangeLog
index 2c4719c..5dc52b1 100644
--- a/bfd/ChangeLog
+++ b/bfd/ChangeLog
@@ -1,3 +1,9 @@
+2014-11-04  Nick Clifton  <nickc@redhat.com>
+
+	PR binutils/17533
+	* archive.c (_bfd_slurp_extended_name_table): Handle archives with
+	corrupt extended name tables.
+
 2013-12-02  Tristan Gingold  <gingold@adacore.com>
 
 	* configure.in: Bump version to 2.24
diff --git a/bfd/archive.c b/bfd/archive.c
index 32b07a7..05a25db 100644
--- a/bfd/archive.c
+++ b/bfd/archive.c
@@ -1292,6 +1292,9 @@ _bfd_slurp_extended_name_table (bfd *abfd)
       amt = namedata->parsed_size;
       if (amt + 1 == 0)
 	goto byebye;
+      /* PR binutils/17533: A corrupt archive can contain an invalid size.  */
+      if (amt > (bfd_size_type) bfd_get_size (abfd))
+	goto byebye;
 
       bfd_ardata (abfd)->extended_names_size = amt;
       bfd_ardata (abfd)->extended_names = (char *) bfd_zalloc (abfd, amt + 1);
@@ -1299,6 +1302,8 @@ _bfd_slurp_extended_name_table (bfd *abfd)
 	{
 	byebye:
 	  free (namedata);
+	  bfd_ardata (abfd)->extended_names = NULL;
+	  bfd_ardata (abfd)->extended_names_size = 0;
 	  return FALSE;
 	}
 
@@ -1307,7 +1312,6 @@ _bfd_slurp_extended_name_table (bfd *abfd)
 	  if (bfd_get_error () != bfd_error_system_call)
 	    bfd_set_error (bfd_error_malformed_archive);
 	  bfd_release (abfd, (bfd_ardata (abfd)->extended_names));
-	  bfd_ardata (abfd)->extended_names = NULL;
 	  goto byebye;
 	}
 
@@ -1315,11 +1319,12 @@ _bfd_slurp_extended_name_table (bfd *abfd)
 	 text, the entries in the list are newline-padded, not null
 	 padded. In SVR4-style archives, the names also have a
 	 trailing '/'.  DOS/NT created archive often have \ in them
-	 We'll fix all problems here..  */
+	 We'll fix all problems here.  */
       {
 	char *ext_names = bfd_ardata (abfd)->extended_names;
 	char *temp = ext_names;
 	char *limit = temp + namedata->parsed_size;
+
 	for (; temp < limit; ++temp)
 	  {
 	    if (*temp == ARFMAG[1])
-- 
2.0.1

